storage/src/main/java/org/apache/kafka/server/log/remote/metadata/storage/ConsumerTask.java

```diff
@@ -64,302 +65,387 @@
-        consumer.assign(assignedMetaTopicPartitions);
+    private void clearResourcesForUnassignedUserTopicPartitions(Set<UserTopicIdPartition> assignedUTPs) {
+        Set<TopicIdPartition> assignedPartitions = assignedUTPs.stream()
+            .map(utp -> utp.topicIdPartition).collect(Collectors.toSet());
+        // Note that there can be previously assigned user-topic-partitions where no records are there to read
+        // (eg) none of the segments for a partition were uploaded. Those partition resources won't be cleared.
+        // It can be fixed later when required since they are empty resources.
+        Set<TopicIdPartition> unassignedPartitions = readOffsetsByUserTopicPartition.keySet()
+            .stream()
+            .filter(e -> !assignedPartitions.contains(e))
+            .collect(Collectors.toSet());
+        unassignedPartitions.forEach(unassignedPartition -> {
+            remotePartitionMetadataEventHandler.clearTopicPartition(unassignedPartition);
+            readOffsetsByUserTopicPartition.remove(unassignedPartition);
+        });
+        log.info("Unassigned user-topic-partitions: {}", unassignedPartitions.size());
+    }
+
+    public void addAssignmentsForPartitions(final Set<TopicIdPartition> partitions) {
+        updateAssignments(Objects.requireNonNull(partitions), Collections.emptySet());
+    }
+
+    public void removeAssignmentsForPartitions(final Set<TopicIdPartition> partitions) {
+        updateAssignments(Collections.emptySet(), Objects.requireNonNull(partitions));
     }
 
-    public void addAssignmentsForPartitions(Set<TopicIdPartition> partitions) {
-        updateAssignmentsForPartitions(partitions, Collections.emptySet());
+    private void updateAssignments(final Set<TopicIdPartition> addedPartitions,
+                                   final Set<TopicIdPartition> removedPartitions) {
+        log.info("Updating assignments for partitions added: {} and removed: {}", addedPartitions, removedPartitions);
+        if (!addedPartitions.isEmpty() || !removedPartitions.isEmpty()) {
+            synchronized (assignPartitionsLock) {
+                final Map<TopicIdPartition, UserTopicIdPartition> idealUserPartitions = new HashMap<>(assignedUserTopicIdPartitions);
+                addedPartitions.forEach(tpId -> idealUserPartitions.putIfAbsent(tpId, newUserTopicIdPartition(tpId)));
+                removedPartitions.forEach(idealUserPartitions::remove);
+                if (!idealUserPartitions.equals(assignedUserTopicIdPartitions)) {
+                    assignedUserTopicIdPartitions = Collections.unmodifiableMap(idealUserPartitions);
+                    isAssignmentChanged = true;
+                }
+                if (isAssignmentChanged) {
+                    log.debug("Assigned user-topic-partitions: {}", assignedUserTopicIdPartitions);
+                    assignPartitionsLock.notifyAll();
+                }
```

Comment:
nit: Why should we add a `if` block after ` isAssignmentChanged = true;`? Can't we just run L291/292 directly like this?
```
                if (!idealUserPartitions.equals(assignedUserTopicIdPartitions)) {
                    assignedUserTopicIdPartitions = Collections.unmodifiableMap(idealUserPartitions);
                    isAssignmentChanged = true;
                    log.debug("Assigned user-topic-partitions: {}", assignedUserTopicIdPartitions);
                    assignPartitionsLock.notifyAll();
                }

```

Reply:
Yes, makes sense

Reply:
Updated